plugins/MumblePluginProcess.hpp

```diff
@@ -0,0 +1,313 @@
+		iovec out;
+		out.iov_base = dst;
+		out.iov_len  = size;
+
+		const auto ret = process_vm_readv(id, &out, 1, &in, 1, 0);
+
+		return (ret != -1 && static_cast< size_t >(ret) == in.iov_len);
+	}
+#endif
+	template< typename T > bool peek(const procptr_t address, T &dst) { return peek(address, &dst, sizeof(T)); }
+
+	template< typename T > T peek(const procptr_t address) {
+		T ret;
+
+		if (!peek(address, &ret, sizeof(T))) {
+			memset(&ret, 0, sizeof(ret));
+		}
+
+		return ret;
+	}
+
+	template< typename T > std::vector< T > peekVector(const procptr_t address, const size_t elements) {
+		try {
+			std::vector< T > var(elements);
+			peek(address, &var[0], sizeof(T) * elements);
+			return var;
+		} catch (std::bad_alloc &) {
+			return std::vector< T >();
+		}
+	}
+
+	/// This function is very useful when there is an array of structures and we only need to store a part of each.
+	/// The Source Engine plugin needs it because there is a structure that happens to be larger in the Linux version of
+	/// Left 4 Dead 2. Since the extra members are unknown, we simply discard the extra bytes when reading the array of
+	/// structures.
+	template< typename T >
+	std::vector< T > peekVector(const procptr_t address, const size_t elements, const size_t realStructSize) {
+		if (realStructSize == sizeof(T)) {
+			// If the structure's size is correct, there is no need for this special function.
+			return peekVector< T >(address, elements);
+		} else if (realStructSize < sizeof(T)) {
+			return std::vector< T >();
+		}
+
+		// Read the full structures into a vector.
+		const auto fullStructs = peekVector< uint8_t >(address, realStructSize * elements);
+
+		// Allocate vector for the known structures.
+		std::vector< T > ret(elements);
+
+		// Copy elements from the vector containing the full structures, discarding extra bytes.
+		auto seek = &fullStructs[0];
+
+		for (size_t i = 0; i < elements; ++i, seek += realStructSize) {
+			memcpy(&ret[i], seek, sizeof(T));
+		}
+
+		return ret;
+	}
+
+	/// Reads the specified amount of data at the specified address and returns it as std::string.
+	/// An empty std::string is returned in case of error.
+	///
+	/// If \p length is 0, the function reads one byte at a time and stops when either '\0' is found or 3 seconds have
+	/// passed. The successfully read data is returned, also in case of error.
+	std::string peekString(const procptr_t address, const size_t length = 0) {
+		std::string string;
+
+		if (length > 0) {
+			string.resize(length);
+
+			if (!peek(address, &string[0], length)) {
+				return std::string();
+			}
+		} else {
+			auto now       = std::chrono::steady_clock::now();
+			const auto end = now + std::chrono::seconds(3);
+
+			for (procptr_t i = 0; now < end; ++i) {
+				char ch = 0;
+				if (!peek(address + i, &ch, sizeof(ch)) || ch == '\0') {
+					break;
+				}
+
+				string += ch;
+
+				// Update current time.
+				now = std::chrono::steady_clock::now();
+			}
+		}
+
+		return string;
+	}
```

Comment:
Implementations like this should go in a separate cpp file

Reply:
It's usually done to prevent the functions from being emitted for every source file that makes use of them.

In the case of plugins it shouldn't matter, as they currently all have a single source file. However:

1. There's no guarantee that plugins will not use more than one source file in the future, especially when the interface you're working on will be available.
2. It tidies up the header quite a bit, especially considering the different function signatures (virtual vs non-virtual) for the two operating systems.

Reply:
Aren't these 2 arguments for moving the implementation into a separate cpp file? :eyes: 

By doing so you obviously have to make all plugins include that source file but in return you get what I would call a cleaner separation between declaration and impementation (which is what is usually expected when working with C/C++)

Reply:
Yes, indeed.